### PR TITLE
Workaround for Typer optional default values with Python calls

### PR DESCRIPTION
**Pull request by rmitsch**
## Goals

Allow calling `init_config_cfg()` from Python with optional arguments not being set. This currently fails due to the behaviour of [Typer](https://github.com/tiangolo/typer) for handling optional arguments in Python calls of functions decorated with `@init_cli.command`. See https://github.com/explosion/spaCy/issues/10727 for a more in-depth description.

## Description

Modified `init_config_cli()` to force default values for optional arguments if they are still of types `typer.models.OptionInfo` or `typer.models.ArgumentInfo`. This seems to be done by Typer automatically on CLI calls, but not on Python calls (see also [the corresponding discussion in the Typer repo](https://github.com/tiangolo/typer/issues/106)).

If such a type is detected, the corresponding variable is set to the configured primitive default value.

### Types of change

Bug fix/enhancement, new test.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Review by svlandeg — COMMENTED** [review]
Thanks for digging into this! I might have a look at the `typer` side of things later this week, too, but either way it's a good idea to make sure our code base is robust against this type of failure.

**Comment by rmitsch** [comment]
Comments/suggestions have been addressed.

**Review by svlandeg — COMMENTED** [review]
Looking into this in a bit more detail, I'm not sure this is the path we want to take. For one, it makes the implementation of the `_cli` function quite ugly and error-prone. Eg. having to think about using `optional_args` instead of the parameters directly -> I could see this leading to errors in the future.

I've started wondering about the original issue as well - I'm not convinced that functions with `_cli` in their name **should** work from a Python call. We definitely don't advertise this API anywhere.

However, the problem remains that even if users call into `init_config` and `save_config` directly, as you'd probably want to recommend them to, then they don't get access to the default values that you'd get from the CLI.

The current implementation does not repeat the default values in e.g. `init_config`, as this also avoids them going out-of-sync across functions. But what if we define some of these default variables as a global on top? Then both functions can access them, they'd always be in sync, and we'd have functions that you can call directly from within Python (though not the `_cli` ones)

**Comment by rmitsch** [comment]
> Looking into this in a bit more detail...
- Ugliness and error proneness of `optional_args`: agreed. The original intention was to not do this for each argument manually. A possible alternative would be to wrap the arguments in a function forcing the default value where necessary:
  ```python
      ...
      config = init_config(
          lang=get_typer_option_value(lang),
         ...
      )
  ```
  It's not much better, but at least avoids a parallel data store for argument values.

- Support of `_cli` calls from Python: it's IMO totally legitimate not to support that. 
- Default values for `init_config()`, `save_config()`: agreed, we shouldn't repeat default values. I'm not a huge fan of setting the default values as globals, but to be fair, I don't have a better idea.

So tl;dr: I think your suggestion of making the default values global is the least bad way to go forward. Should we bundle them in an object so they don't clutter the module too much, like this?
```python
class InitDefaultValues:
    output_file = "-"
    pretraining = False
    ...
    
```

**Comment by rmitsch** [comment]
Removed forcing of optional default values, added synchronized default values for `init_config_cli()` and `init_cli()`.

File: spacy/cli/init_config.py [hunk 1]
```diff
@@ -44,6 +59,7 @@ def init_config_cli(
 
     DOCS: https://spacy.io/api/cli#init-config
     """
+
```
Comment (adrianeboyd) [inline spacy/cli/init_config.py:62]: I don't think these newlines were intentional or added by black?

File: spacy/cli/init_config.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 from typing import Optional, List, Tuple
 from enum import Enum
 from pathlib import Path
+
```
Comment (svlandeg) [inline spacy/cli/init_config.py:4]:
```suggestion
```

File: spacy/cli/init_config.py [hunk 3]
```diff
@@ -52,6 +67,7 @@ def init_config_cli(
             "The provided output file already exists. To force overwriting the config file, set the --force or -F flag.",
             exits=1,
         )
+
```
Comment (svlandeg) [inline spacy/cli/init_config.py:70]:
```suggestion
```

File: spacy/tests/test_cli.py [hunk 4]
```diff
@@ -750,6 +751,28 @@ def test_debug_data_compile_gold():
     assert data["boundary_cross_ents"] == 1
 
 
+@pytest.mark.issue(10727)
+def test_init_config_from_python_without_optional_args():
```
Comment (svlandeg) [inline spacy/tests/test_cli.py:755]: I'm not sure we need this test? It ensures that an `AttributeError` is thrown when `init_config_cli` is called, but this isn't really the behaviour we're targeting, it's more like an unfortunate side effect we can work around by calling the other internal methods that now do have defaults? So maybe we should check the latter instead?

**Review by svlandeg — COMMENTED** [review]
Looking good! Removed some of the introduced newlines and just had one more question on the test.

File: spacy/tests/test_cli.py (hunk 4 above)
Comment (rmitsch) [reply spacy/tests/test_cli.py:755]: True, that's a leftover. What would we check for `init_config()` though? Seems to me like the final change of adding default params is so trivial that we can forego the new test completely and rely on the existing ones.

File: spacy/tests/test_cli.py [hunk 5]
```diff
@@ -23,6 +23,7 @@
 from spacy.cli.debug_data import _get_spans_length_freq_dist
 from spacy.cli.download import get_compatibility, get_version
 from spacy.cli.init_config import RECOMMENDATIONS, init_config, fill_config
+from spacy.cli.init_config import init_config_cli, Optimizations
```
Comment (svlandeg) [inline spacy/tests/test_cli.py:26]:
```suggestion
```

File: spacy/cli/init_config.py [hunk 6]
```diff
@@ -12,7 +12,7 @@
 from ..schemas import RecommendationSchema
```
Comment (svlandeg) [inline spacy/cli/init_config.py:12]:
```suggestion
from ..schemas import RecommendationSchema
from ..util import SimpleFrozenList
```

File: spacy/cli/init_config.py [hunk 7]
```diff
@@ -12,7 +12,7 @@
...
 from ._util import init_cli, Arg, Opt, show_validation_error, COMMAND
 from ._util import string_to_list, import_code
```
Comment (svlandeg) [inline spacy/cli/init_config.py:14]:
```suggestion
from ._util import string_to_list, import_code

```

File: spacy/cli/init_config.py [hunk 8]
```diff
@@ -10,9 +10,11 @@
 from .. import util
 from ..language import DEFAULT_CONFIG_PRETRAIN_PATH
 from ..schemas import RecommendationSchema
+from ..util import SimpleFrozenList
 from ._util import init_cli, Arg, Opt, show_validation_error, COMMAND
 from ._util import string_to_list, import_code
 
+from ..util import SimpleFrozenList
```
Comment (svlandeg) [inline spacy/cli/init_config.py:17]:
```suggestion
```

**Review by svlandeg — APPROVED** [review]
Should be good to merge once the CI goes green!

**Comment by rmitsch** [comment]
@svlandeg I think we can merge this?
